src/optimagic/logging/logger.py

```diff
@@ -0,0 +1,320 @@
+    CriterionEvaluationWithId,
+    ExistenceStrategy,
+    ExistenceStrategyLiteral,
+    IterationHistory,
+    MultiStartIterationHistory,
+    ProblemInitialization,
+    ProblemInitializationWithId,
+    StepResult,
+    StepResultWithId,
+    StepType,
+)
+from optimagic.parameters.tree_registry import get_registry
+from optimagic.typing import OptimizationType, OptimizationTypeLiteral, PyTree
+
+
+class Logger:
+    """A logger class that manages and retrieves optimization and exploration data.
+
+    This class handles storing and retrieving iterations, steps, and problem
+    initialization data using various stores.
+    It provides methods to read iteration history, retrieve
+    specific iterations, and handle multistart optimization history.
+
+    Args:
+        iteration_store: A non-updatable store for iteration data.
+        step_store: An updatable store for step data.
+        problem_store: An updatable store for problem initialization data.
+
+    """
+
+    def __init__(
+        self,
+        iteration_store: NonUpdatableKeyValueStore[
+            CriterionEvaluationResult, CriterionEvaluationWithId
+        ],
+        step_store: UpdatableKeyValueStore[StepResult, StepResultWithId],
+        problem_store: UpdatableKeyValueStore[
+            ProblemInitialization, ProblemInitializationWithId
+        ],
+    ):
+        self.step_store = step_store
+        self.iteration_store = iteration_store
+        self.problem_store = problem_store
+        self._pytree_registry = get_registry(extended=True)
+
+    @cached_property
+    def _treedef(self) -> tuple[list[Any], Any]:
+        return tree_flatten(self.read_start_params(), registry=self._pytree_registry)[
+            -1
+        ]
+
+    def read_iteration(self, iteration: int) -> CriterionEvaluationWithId:
+        """Read a specific iteration from the iteration store.
+
+        Args:
+            iteration: The iteration number to read. Negative values read from the end.
+
+        Returns:
+            A `CriterionEvaluationWithId` object containing the iteration data.
+
+        Raises:
+            IndexError: If the iteration is invalid or the store is empty.
+
+        """
+        if iteration >= 0:
+            rowid = iteration + 1
+        else:
+            try:
+                last_row = self.iteration_store.select_last_rows(1)
+                highest_rowid = last_row[0].rowid
+            except IndexError as e:
+                raise IndexError(
+                    "Invalid iteration request, iteration store is empty"
+                ) from e
+
+            # iteration is negative here!
+            assert highest_rowid is not None
+            rowid = highest_rowid + iteration + 1
+
+        row_list = self.iteration_store.select(rowid)
+
+        if len(row_list) == 0:
+            raise IndexError(f"Invalid iteration requested: {iteration}")
+        else:
+            data = row_list[0]
+
+        params = tree_unflatten(
```

Comment:
I think there is actually not good reason for storing the parameters in a flat format. 

We define `external_x` (i.e. flattened but not reparametrized params) [here](https://github.com/OpenSourceEconomics/optimagic/blob/354155d38157c861ae8988e45301237a577096d1/src/optimagic/optimization/internal_criterion_template.py#L101) and then just use it for logging. But we also have a non-flat `current_params` defined. If we write that into the database, we can get rid of the unflattening here. 

There might have been a reason in the dashboard code but that reason is definitely no longer valid.

Reply:
Agree, will do